### ISSUE TITLE
Cannot load MaskRCNN checkpoint

### ISSUE THREAD

**Issue by djdongjin** [tensorflow/models#8789]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x] I am using the latest TensorFlow Model Garden release and TensorFlow 2.
- [x] I am reporting the issue to the correct repository. (Model Garden official or research directory)
- [x] I checked to make sure that this issue has not been filed already.

## 1. The entire URL of the file you are using

https://github.com/tensorflow/models/tree/master/official/vision/detection/main.py

## 2. Describe the bug

In the official implementation of the MaskRCNN model, the evaluation command cannot correctly load the checkpoint saved by the training command.

## 3. Steps to reproduce

```python
export DATA_DIR=${STORAGE_BUCKET}/coco

export PYTHONPATH="${PYTHONPATH}:${HOME}/models"
export MODEL_DIR=${STORAGE_BUCKET}/mask-rcnn
export RESNET_CHECKPOINT=${STORAGE_BUCKET}/resnet50-2018-02-07
export TRAIN_FILE_PATTERN=${DATA_DIR}/train-*
export EVAL_FILE_PATTERN=${DATA_DIR}/val-*
export VAL_JSON_FILE=${DATA_DIR}/instances_val2017.json

python3 official/vision/detection/main.py \
        --strategy_type=tpu \
        --tpu=${TPU_NAME} \
        --model_dir=${MODEL_DIR} \
        --mode="train" \
        --model=mask_rcnn \
        --params_override="{train: { checkpoint: { path: ${RESNET_CHECKPOINT}, prefix: resnet50/ }, train_file_pattern: ${TRAIN_FILE_PATTERN} }, eval: { val_json_file: ${VAL_JSON_FILE}, eval_file_pattern: ${EVAL_FILE_PATTERN}, eval_samples: 5000} }"

python3 official/vision/detection/main.py \
        --strategy_type=tpu \
      --tpu=${TPU_NAME} \
      --model_dir=${MODEL_DIR} \
      --mode="eval" \
      --model=mask_rcnn \
      --params_override="{eval: { val_json_file: ${VAL_JSON_FILE}, eval_file_pattern: ${EVAL_FILE_PATTERN}, eval_samples: 5000 } }"
```
```python
# Error message
.......
W0706 01:35:19.699204 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-122.kernel
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-122.bias
W0706 01:35:19.699246 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-122.bias
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-123.kernel
W0706 01:35:19.699290 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-123.kernel
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-123.bias
W0706 01:35:19.699333 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-123.bias
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-124.kernel
W0706 01:35:19.699375 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-124.kernel
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-124.bias
W0706 01:35:19.699415 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-124.bias
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-125.kernel
W0706 01:35:19.699458 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-125.kernel
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-125.bias
W0706 01:35:19.699500 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-125.bias
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-126.kernel
W0706 01:35:19.699541 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-126.kernel
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-126.bias
W0706 01:35:19.699584 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-126.bias
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-127.kernel                                                                        W0706 01:35:19.699635 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-127.kernel
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-127.bias
W0706 01:35:19.699677 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-127.bias
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-128.kernel
W0706 01:35:19.699716 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-128.kernel
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-128.bias
W0706 01:35:19.699758 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-128.bias
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-129.kernel
W0706 01:35:19.699801 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-129.kernel
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-129.bias
W0706 01:35:19.699842 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-129.bias
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-130.kernel
W0706 01:35:19.699888 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-130.kernel
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-130.bias
W0706 01:35:19.699936 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-130.bias
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-131.kernel
W0706 01:35:19.700003 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-131.kernel
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-131.bias
W0706 01:35:19.700045 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-131.bias
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-132.kernel
W0706 01:35:19.700086 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-132.kernel
WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-132.bias
W0706 01:35:19.700129 139854688462656 util.py:150] Unresolved object in checkpoint: (root).optimizer's state 'momentum' for (root).model.layer_with_weights-132.bias
WARNING:tensorflow:A checkpoint was restored (e.g. tf.train.Checkpoint.restore or tf.keras.Model.load_weights) but not all checkpointed values were used. See above for specific issues. Use expect_partial()
on the load status object, e.g. tf.train.Checkpoint.restore(...).expect_partial(), to silence these warnings, or use assert_consumed() to make the check explicit. See https://www.tensorflow.org/guide/checkp
oint#loading_mechanics for details.
W0706 01:35:19.700174 139854688462656 util.py:158] A checkpoint was restored (e.g. tf.train.Checkpoint.restore or tf.keras.Model.load_weights) but not all checkpointed values were used. See above for specif
ic issues. Use expect_partial() on the load status object, e.g. tf.train.Checkpoint.restore(...).expect_partial(), to silence these warnings, or use assert_consumed() to make the check explicit. See https:/
/www.tensorflow.org/guide/checkpoint#loading_mechanics for details.
```

**Comment by peaceiris** [comment]
We also have a similar problem with the `object_detection`.

**Comment by gagika** [comment]
It looks like you are training and evaluating on a TPU. 
Please make sure both GCP VM and TPU have Tensorflow 2.2 (or tf-nightly). As there are some known issues of running Mask-RCNN in Tensorflow 2.1.

Also not sure if you are using a correct pretrained ResNet50 checkpoint, you can use this one: 
```
export RESNET_CHECKPOINT=gs://cloud-tpu-checkpoints/retinanet/resnet50-checkpoint-2018-02-07
```

Please follow a detailed [Mask-RCNN TPU tutorial](https://cloud.google.com/tpu/docs/tutorials/mask-rcnn-2.x) and let us know if any issues remain.

**Comment by djdongjin** [comment]
Yes I am using a TPU and following the mentioned tutorial and using tf-nightly. The Resnet checkpoint I used is the one given in official/detection/README.md (https://storage.cloud.google.com/cloud-tpu-checkpoints/model-garden-vision/detection/resnet50-2018-02-07.tar.gz). I downloaded the checkpoint and uploaded to my gc bucket.

**Comment by gagika** [comment]
Are you creating both compute VM and TPU with `ctpu up` command with an argument `--tf-version=nightly`?
It's important to make sure both VM and TPU have the same Tensorflow versions.

**Comment by djdongjin** [comment]
Yes, I created them with `ctpu --tf-version=nightly` argument. Let me know if you can run it correctly. I can also rerun it later if possible and update here.

**Comment by gagika** [comment]
I looked into the issue, Mask-RCNN model structure is little bit different when it's built for training vs evaluation. It looks like the checkpoint from training has more values than evaluation, that is what these warnings are saying . Before printing this warnings it prints box AP and mask AP numbers, and it gets to the same accuracy as before (when training for 22500 steps, box AP is 0.37+ and Mask AP is 0.33+) 

Before we fix the issue, you can silence these warnings by changing restore checkpoint in `evaluate_checkpoint` function of [distributed_executor.py](https://github.com/tensorflow/models/blob/master/official/modeling/training/distributed_executor.py#L701) with:

```
status = checkpoint.restore(checkpoint_path)
status.expect_partial().assert_existing_objects_matched()
```
